client/state/data-layer/wpcom-http/pipeline/retry-on-failure/test/index.js

```diff
@@ -0,0 +1,184 @@
+/**
+ * External dependencies
+ */
+import deepFreeze from 'deep-freeze';
+import { expect } from 'chai';
+import { spy } from 'sinon';
+import { merge } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { useFakeTimers } from 'test/helpers/use-sinon';
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { retryOnFailure as rof } from '../';
+import { noRetry, simpleRetry, exponentialBackoff } from '../policies';
+
+const retryOnFailure = rof();
+const retryWithDelay = delay => rof( { getDelay: () => delay } );
+
+const nextError = { fail: 'failed big time' };
+
+const getSites = deepFreeze( http( {
+	method: 'GET',
+	path: '/sites',
+	apiVersion: 'v1',
+} ) );
+
+const withRetries = retryCount => actionOrInbound =>
+	undefined !== actionOrInbound.originalRequest
+		? merge( actionOrInbound, { originalRequest: withRetries( retryCount )( actionOrInbound.originalRequest ) } )
+		: merge( actionOrInbound, { meta: { dataLayer: { retryCount } } } );
+
+describe( '#retryOnFailure', () => {
+	let clock;
+	let dispatch;
+	let store;
+
+	useFakeTimers( fakeClock => clock = fakeClock );
+
+	beforeEach( () => {
+		dispatch = spy();
+		store = { dispatch };
+	} );
+
+	it( 'should pass through initially successful requests', () => {
+		const inbound = { nextData: 1, originalRequest: getSites, store };
+
+		expect( retryOnFailure( inbound ) ).to.equal( inbound );
+
+		clock.tick( 20000 );
+		expect( dispatch ).to.have.not.been.called;
+	} );
+
+	it( 'should pass through no-retry failed requests', () => {
+		const originalRequest = { ...getSites, options: { whenFailing: noRetry() } };
+		const inbound = { nextError, originalRequest, store };
+
+		expect( retryOnFailure( inbound ) ).to.equal( inbound );
+
+		clock.tick( 20000 );
+		expect( dispatch ).to.have.not.been.called;
+	} );
+
+	it( 'should pass through POST requests', () => {
+		const originalRequest = { ...getSites, method: 'POST', options: { whenFailing: simpleRetry( { delay: 1000 } ) } };
+		const inbound = { nextError, originalRequest, store };
+
+		expect( retryOnFailure( inbound ) ).to.equal( inbound );
+
+		clock.tick( 20000 );
+		expect( dispatch ).to.have.not.been.called;
+	} );
+
+	it( 'should requeue a plain failed request', () => {
+		const inbound = { nextError, originalRequest: getSites, store };
+
+		expect( retryWithDelay( 1337 )( inbound ) ).to.have.property( 'shouldAbort', true );
+		expect( dispatch ).to.have.not.been.called;
+
+		clock.tick( 1337 );
+		expect( dispatch ).to.have.been.calledWith( withRetries( 1 )( getSites ) );
+	} );
+
+	it( 'should requeue only up to `maxAttempts`', () => {
+		const originalRequest = { ...getSites, options: { whenFailing: simpleRetry( { delay: 1000, maxAttempts: 3 } ) } };
+		const inbound = { nextError, originalRequest, store };
+		const retryIt = retryWithDelay( 1337 );
+
+		// retry 1
```

Comment:
a retry, as indicated in the comment. it's all very confusing, but the fact that we have a `nextError` attached indicates that it's a response